Add entry IDs to the memory queue

faec opened the pull request:
## What does this PR do?

Adds an incrementing ID to events in the memory queue, as described in https://github.com/elastic/elastic-agent-shipper/issues/27. This is needed by the shipper, to report to inputs whether their events have been sent yet.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

Related: https://github.com/elastic/elastic-agent-shipper/issues/27

elasticmachine commented:
Pinging @elastic/elastic-agent-data-plane  (Team:Elastic-Agent-Data-Plane)

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-32541/detail/PR-32541/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-32541/detail/PR-32541/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-32541/detail/PR-32541/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-32541/detail/PR-32541/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_32541.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-07-28T20:49:20.500Z&rangeTo=2022-07-28T21:09:20.500Z&transactionName=Beats/beats/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=25a1eef3c4caf21e38225cac366cee4a&transactionId=f4c16bf4f555ddac)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-07-28T20:59:20.500+0000


* Duration: 21 min 11 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 3  |
| Skipped      | 0 |
| Total        | 3   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![1](https://img.shields.io/badge/1%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-32541/detail/PR-32541/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/PR-32541/runs/1/steps/6092/log/?start=0">here</a></li>
<li>Description: <code>Error 'java.io.NotSerializableException: java.util.regex.Matcher'</code></l1>
</ul>
  
  </p>
  </details>



## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

rdner commented on libbeat/publisher/queue/memqueue/ringbuf.go:
```diff
@@ -52,10 +52,10 @@ type region struct {
+/*type producerEventState struct {
+	eventIndex uint64 // Order of the event in its originating producer, independent of queue ID
+	producer   *ackProducer
+}*/
```
Is it supposed to be commented?

faec replied:
Nope, it's supposed to be deleted :-)

rdner commented on libbeat/publisher/queue/diskqueue/consumer.go:
```diff
@@ -86,10 +86,14 @@ func (batch *diskQueueBatch) Count() int {
 }
 
+func (batch *diskQueueBatch) ID(i int) queue.EntryID {
+	return 0
```
So the disk queue would not support the incremental IDs? Or it's going to be implemented in a separate PR?

faec replied:
Yes, this is just for the memory queue, disk queue will be another PR (the shipper doesn't support the disk queue yet although with Lee's pending PR that should change any day now)

rdner commented on libbeat/publisher/queue/memqueue/broker.go:
```diff
@@ -380,10 +388,14 @@ func (b *batch) Count() int {
 }
 
-func (b *batch) Event(i int) interface{} {
+func (b *batch) Entry(i int) interface{} {
 	return b.entries[i].event
 }
 
+func (b *batch) ID(i int) queue.EntryID {
+	return b.entries[i].id
+}
+
```
Don't we need a boundary check and a returned error in exported functions like this? It should not panic if the `i` is out of range, I think?

rdner commented on libbeat/publisher/queue/memqueue/eventloop.go:
```diff
@@ -143,16 +145,18 @@ func (l *directEventLoop) handleMetricsRequest(req *metricsRequest) {
-			seq:   req.seq,
-			state: st,
-		})
+		if req.resp != nil {
```
I think the comment to this `insert` function is outdated since it does not actually return anything and I think it needs a comment inside the function to explain why `req.resp` can be `nil` and how we handle it.

faec replied:
`req.resp` actually can't be nil! That check is leftover from an earlier iteration where it could be, so I removed it.

rdner commented on libbeat/publisher/queue/memqueue/eventloop.go:
```diff
@@ -203,49 +207,25 @@ func (l *directEventLoop) processACK(lst chanList, N int) {
-			// log.Debug("seq number already acked: ", st.seq)
-
-			client.state = nil
+		idx := (firstIndex + i) % len(entries)
```
It's not obvious what's going on with indexing here, could you please add a comment explaining this?

Also, I don't see where `len(entries)` is changed in the loop, it might worth to just pre-calculate this into `i` when the loop starts.

faec replied:
`len(entries)` doesn't change, only `i` does. I've added more comments clarifying where this value is coming from, see if that makes more sense.

rdner commented on libbeat/publisher/queue/memqueue/eventloop.go:
```diff
@@ -203,49 +207,25 @@ func (l *directEventLoop) processACK(lst chanList, N int) {
+		entry := &entries[idx]
+
+		producer := entry.producer
+		entry.producer = nil
```
Is this how we mark that we have visited this entry before?

faec replied:
Yes (though it may already be nil if the producer doesn't accept ACK callbacks). I've added a comment explaining.

rdner commented on libbeat/publisher/queue/memqueue/produce.go:
```diff
@@ -42,66 +41,83 @@ type openState struct {
+func (p *ackProducer) Publish(event interface{}) (queue.EntryID, bool) {
+	id, published := p.openState.publish(p.makePushRequest(event))
+	if published {
+		p.producedCount++
```
Is it thread safe to just increment the counter in both `Publish` and `TryPublish`?

faec replied:
No, but neither is the code that this replaces. The queues enforce thread-safe access but a single producer is only supposed to be used by a single goroutine at a time (though now that you mention it I'm not sure the current shipper code respects that).

faec replied:
(Thinking more: I think the shipper case is actually ok, because it doesn't use or need a producer callback, so it won't hit `ackProducer`, just `forgetfulProducer` which doesn't need to track anything.)

faec commented:
Added some more bits around metrics reporting, since I realized this code reported event IDs to the producers and consumers but the shipper also needs a standalone call to get the current ID, which I've added to the metrics struct already provided by the memory queue.

faec commented:
The existing tests pass, the new tests for the event ID feature still need work:
- they involve a `time.Sleep` which is undesirable. Not sure yet if I can remove it, though, since the memory queue propagates ACKs asynchronously.
- some of them don't pass yet, particularly with the buffered version. I'm debugging that now, not 100% sure yet whether the problem is in the implementation or tests.

cmacknz commented on libbeat/publisher/queue/memqueue/produce.go:
```diff
@@ -42,66 +41,83 @@ type openState struct {
+// event acknowledgement callbacks can be coalesced into a single call.
+// It is defined as an explicit type to reduce cross-confusion with the id
+// of an event in the queue itself, which is a queue.EntryID.
+// with the
```
```go
// with the
```

The comment here isn't finished

cmacknz commented on libbeat/publisher/queue/memqueue/produce.go:
```diff
@@ -42,66 +41,83 @@ type openState struct {
-func (p *forgetfulProducer) Publish(event interface{}) bool {
-	return p.openState.publish(pushRequest{event: event})
+func (p *forgetfulProducer) makePushRequest(event interface{}) pushRequest {
+	resp := make(chan queue.EntryID, 1)
```
We make a channel for each event? Is it worth making one for each batch? Could we make a version of `Publish` that accepts an entire event batch instead of an individual event?

faec replied:
It's not impossible but it would significantly complicate things: right now the memory queue accepts publish requests iff it has at least one free event slot, because then it can definitely handle the request without blocking. Accepting multiple events per push request would mean that's no longer true, so we'd need to assemble the response over time, and both event loops would need more complicated state handling to spread the response across multiple iterations.

However, ephemeral channels are quite cheap and this pattern is already used heavily in both queues. (We could, though, probably cache a single response channel for each producer, since the producers themselves are only supposed to be used from a single goroutine?)

cmacknz replied:
Thanks, the pattern from the shipper side is to try to publish entire batches instead of individual events. I was more interested in seeing if we could improve the API for that use case, but it sounds like it isn't worth it unless there is some evidence of a performance issue.

faec replied:
Ah, in that case yes I'd say it's a possible future optimization but probably not worth the complexity until we see where the real bottlenecks are. (My current belief is that the channeled API calls are negligible compared to the actual queueing process, so our current approach of calling single-event publish in a for loop doesn't lose much. It also gives us free semi-fair balancing between different inputs -- in the simplest implementation, keeping published batches together for the queue call would block all other inputs while a large publish request was being processed.)

cmacknz commented on libbeat/publisher/queue/memqueue/eventloop.go:
```diff
@@ -203,49 +216,34 @@ func (l *directEventLoop) processACK(lst chanList, N int) {
+	// We want to acknowledge N events starting at position firstIndex
+	// in the entries array.
+	// We iterate over the events from last to first, so we encounter the
+	// highest produer IDs first and can skip subsequent callbacks to the
```
```suggestion
	// highest producer IDs first and can skip subsequent callbacks to the
```

cmacknz commented on libbeat/publisher/queue/memqueue/eventloop.go:
```diff
@@ -323,11 +321,28 @@ func (l *bufferingEventLoop) run() {
-	if l.insert(req) {
+	if l.insert(req, l.nextEntryID) {
+		// Send back the new event id.
+		req.resp <- l.nextEntryID
```
With these response channels, is there any chance that the receiving side has been cancelled or is otherwise unresponsive? Do we need to handle cases where this send blocks unexpectedly?

faec replied:
No (though I could add more comments):
- these channels are only accessible internally to the queue
- any channel-based API call immediately listens on the response channel once the request is sent
- the response channels have buffer size 1 so the response can't block the main queue goroutine.
- once the queue accepts a request it will always send a response

cmacknz commented on libbeat/publisher/queue/memqueue/queue_test.go:
```diff
@@ -175,3 +175,98 @@ func TestAdjustInputQueueSize(t *testing.T) {
+			assert.Equal(t, batch.Count(), 1, "Returned batch should have 1 entry")
+			assert.Equal(t, batch.ID(0), queue.EntryID(i), "Consumed entry IDs should be ordered the same as when they were produced")
+
+			time.Sleep(100 * time.Millisecond)
```
What is this delay for?

faec replied:
troubleshooting, it isn't really needed and I'll remove it once I get the last test passing

cmacknz commented on libbeat/publisher/queue/memqueue/queue_test.go:
```diff
@@ -175,3 +176,165 @@ func TestAdjustInputQueueSize(t *testing.T) {
+
+	testForward := func(q queue.Queue) {
+		waiter := &producerACKWaiter{}
+		producer := q.Producer(queue.ProducerConfig{ACK: waiter.ack})
```
Is it worth having a test case for concurrent producers?

cmacknz commented:
LGTM, I'll wait for @rdner to confirm that his comments have been addressed before approving.

cmacknz commented:
Would also be good to get @leehinman to look at this, I've pinged him on Slack.

leehinman commented on libbeat/publisher/queue/diskqueue/consumer.go:
```diff
@@ -86,10 +86,14 @@ func (batch *diskQueueBatch) Count() int {
 }
 
+func (batch *diskQueueBatch) ID(i int) queue.EntryID {
+	return 0
```
Am I correct in assuming a "real" event ids for the disk queue would be in a follow on PR?  If so can we open an issue and make sure it is marked as "related" in this PR?

faec replied:
Yes, the issue is https://github.com/elastic/elastic-agent-shipper/issues/27

leehinman reviewed (approved):
In a follow on PR could we get a doc with some high level design docs/diagrams.
